### PR TITLE
Add PeerStore content & TEMP generalize PeerID guide

### PR DESCRIPTION
**Pull request by DannyS03**
## Context

- Generalizes the `PeerID` guide to add content for `PeerStore`, but to reconsider structure based on the proposed structure changes. Could also consider creating a section and a dedicated index about peer-specific content
- Part of #33, #35

## Latest preview

Please view the latest Fleek preview [here](https://bafybeibizxqqedycmfmmtn4r6o3xtmkr3h6o5hmaajsskxws4eeh4do2xi.on.fleek.co/concepts/peers/).

### CODE REVIEW

File: content/concepts/peers.md [hunk 1]
```diff
@@ -1,16 +1,16 @@
 ---
-title: Peer Identity
+title: All about Peers
 weight: 4
 ---
 
+### What is a PeerId
+
 A Peer Identity (often written `PeerId`) is a unique reference to a specific
```
Comment (marten-seemann) [inline content/concepts/peers.md:8]: Can we just write them as Peer ID? There's no need to pretend that it's a variable.
(I'm fine if that's done in a separate PR).

File: content/concepts/peers.md [hunk 2]
```diff
@@ -107,3 +106,35 @@ their addresses using [peer routing](/concepts/peer-routing/).
 [definition_multihash]: /reference/glossary/#multihash
 
 [spec_peerid]: https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md
+
+## PeerStore
```
Comment (marten-seemann) [inline content/concepts/peers.md:110]: This seems overly specific to the Go implementation. Would it be sufficient to just say that a node will typically have a temporary store to store peer keys, addresses and associated metadata.

File: content/concepts/peers.md [hunk 3]
```diff
@@ -107,3 +106,35 @@ their addresses using [peer routing](/concepts/peer-routing/).
...
+
+The `PeerStore` also uses an Event Emitter to notify interested parties of relevant events, such as peer discovery.
+
+### Discovery events
+
+A discovery method is likely needed if a peer is undiscoverable using the `PeerStore`. A peer `multiaddr` is typically discovered with their `PeerId`. Once the network successfully discovers a peer `multiaddr`, the peer discovery protocol will emit a peer event to add the `PeerInfo` and peer `multiaddr`  to the `PeerStore`. Learn more about how to discover un-{known, identified} peers on the Peer Routing guide. 
+
+This is one way that the network updates the `PeerStore`. In general, an identify protocol automatically runs on every connection when multiplexing is enabled. The protocol will put the `multiaddrs` and protocols identifiers provided by the peer to the `PeerStore`. Similarly, an `IdentifyPush` protocol waits for change notifications about protocols that a peer supports and updates the `PeerStore` accordingly.
+
+### Retreival events
```
Comment (marten-seemann) [inline content/concepts/peers.md:134]:
```suggestion
### Retrieval events
```

File: content/concepts/peers.md [hunk 4]
```diff
@@ -107,3 +106,35 @@ their addresses using [peer routing](/concepts/peer-routing/).
...
+
+The `PeerStore` also provides an API for the components of the inner book, as well as data events.
+
+A `datastore` helps with data persistence for peers that may have been offline or reset, to improve connection efficiency on the libp2p network. A libp2p node will need to receive a `datastore` to persist data across restarts—a `datastore` stores data as key-value pairs. The store maintains data persistence and connection efficiency by not constantly updating the `datastore` with new data. Instead, the `datastore` stores new data only after reaching a certain threshold of peers out-of-date, and when a node stops to, batch writes to the datastore.
+
+The `PeerID` appends the `datastore` key for each data namespace. The namespaces were defined as follows:
+
+The `PeerStore` also uses an Event Emitter to notify interested parties of relevant events, such as peer discovery.
+
+### Discovery events
```
Comment (marten-seemann) [inline content/concepts/peers.md:128]:
```suggestion
### Peer Discovery
```

File: content/concepts/peers.md [hunk 5]
```diff
@@ -107,3 +106,35 @@ their addresses using [peer routing](/concepts/peer-routing/).
...
+
+A discovery method is likely needed if a peer is undiscoverable using the `PeerStore`. A peer `multiaddr` is typically discovered with their `PeerId`. Once the network successfully discovers a peer `multiaddr`, the peer discovery protocol will emit a peer event to add the `PeerInfo` and peer `multiaddr`  to the `PeerStore`. Learn more about how to discover un-{known, identified} peers on the Peer Routing guide.
```
Comment (marten-seemann) [inline content/concepts/peers.md:130]: This is also very implementation-specific. For example, go-libp2p doesn't use events.

File: content/concepts/peers.md [hunk 6]
```diff
@@ -107,3 +106,35 @@ their addresses using [peer routing](/concepts/peer-routing/).
...
+
+The `PeerID` appends the `datastore` key for each data namespace. The namespaces were defined as follows:
+
+The `PeerStore` also uses an Event Emitter to notify interested parties of relevant events, such as peer discovery.
+
+### Discovery events
+
+A discovery method is likely needed if a peer is undiscoverable using the `PeerStore`. A peer `multiaddr` is typically discovered with their `PeerId`. Once the network successfully discovers a peer `multiaddr`, the peer discovery protocol will emit a peer event to add the `PeerInfo` and peer `multiaddr`  to the `PeerStore`. Learn more about how to discover un-{known, identified} peers on the Peer Routing guide. 
+
+This is one way that the network updates the `PeerStore`. In general, an identify protocol automatically runs on every connection when multiplexing is enabled. The protocol will put the `multiaddrs` and protocols identifiers provided by the peer to the `PeerStore`. Similarly, an `IdentifyPush` protocol waits for change notifications about protocols that a peer supports and updates the `PeerStore` accordingly.
```
Comment (marten-seemann) [inline content/concepts/peers.md:132]: It should be "the Identify protocol".

Can we link to the doc describing Identify?

File: content/concepts/peers.md (hunk 1 above)
Comment (DannyS03) [reply content/concepts/peers.md:8]: Yes

File: content/concepts/peers.md (hunk 5 above)
Comment (DannyS03) [reply content/concepts/peers.md]: Updated

File: content/concepts/peers.md (hunk 4 above)
Comment (DannyS03) [reply content/concepts/peers.md]: Updated

File: content/concepts/peers.md (hunk 3 above)
Comment (DannyS03) [reply content/concepts/peers.md]: Updated

File: content/concepts/peers.md (hunk 2 above)
Comment (DannyS03) [reply content/concepts/peers.md]: Updated

File: content/concepts/peers.md (hunk 6 above)
Comment (DannyS03) [reply content/concepts/peers.md]: Updated

**Comment by DannyS03** [comment]
Thanks @marten-seemann! Added changes

File: content/concepts/peers.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+In general, an [Identify protocol][identity] automatically runs on every connection when 
+multiplexing is enabled. The protocol will put the multiaddrs and protocols identifiers provided 
+by the peer to the Peer Store. Similarly, the Identity protocol waits for change notifications 
+about protocols that a peer supports and updates the Peer Store accordingly.
+
+### Peer Retrieval
+
+The Peer Store notifies the libp2p network when the network discovers 
+a new peer. The Peer Store also notifies the network about changes to the Peer Info
+about the peer, such as the peer's supported protocols and known multiaddrs.
```
[the remark this answers is not in the document]
Comment (DannyS03) [reply content/concepts/peers.md]: Removed

File: content/concepts/peers.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+that each peer supports, which may change over time, which the `protocolBook` accounts 
+for; a `metadataBook` keeps track of the available peer metadata, which is stored in a 
+key-value fashion, where a key identifier (string) represents a metadata value (Uint8Array). 
+There is also an API to expose the components of the inner book, as well as data events 
+to emit data about new information and changes.
+  
+{{% /notice %}}
+
+A datastore helps with data persistence for peers that may have been offline or reset, to 
+improve connection efficiency on the libp2p network. A libp2p node will need to receive a
```
[the remark this answers is not in the document]
Comment (DannyS03) [reply content/concepts/peers.md]: hmm, I see. Suggestion incorporated

File: content/concepts/peers.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+addresses and associated metadata. `PeerStore`, also known as a `PeerBook`, 
+is a register in libp2p that holds an updated data (Peer Info) registry of all 
+known peers. Other peers can dial the peer store and listen for updates and learn 
+about any peer within the network. The peer store works like a phone or address book; 
+think of it like a universal multiaddr book that maintains the source of truth for all
+known peers.
+
+{{% notice "note" %}}
+
+With different design choices possible, here is a snapshot of how the js-libp2p
```
[the remark this answers is not in the document]
Comment (DannyS03) [reply content/concepts/peers.md]: Removed

File: content/concepts/peers.md [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+Peer Info combines a Peer Id with a set of [multiaddrs][definition_multiaddr]
+that the peer is listening on.
+
+## Peer Store
+
+A libp2p node will typically have a temporary store to store peer keys, 
+addresses and associated metadata. `PeerStore`, also known as a `PeerBook`, 
+is a register in libp2p that holds an updated data (Peer Info) registry of all 
+known peers. Other peers can dial the peer store and listen for updates and learn
```
[the remark this answers is not in the document]
Comment (DannyS03) [reply content/concepts/peers.md]: Removed

File: content/concepts/peers.md [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+Another common libp2p data structure related to peer identity is the `PeerInfo`
+structure.
+
+Peer Info combines a Peer Id with a set of [multiaddrs][definition_multiaddr]
+that the peer is listening on.
+
+## Peer Store
+
+A libp2p node will typically have a temporary store to store peer keys, 
+addresses and associated metadata. `PeerStore`, also known as a `PeerBook`,
```
[the remark this answers is not in the document]
Comment (DannyS03) [reply content/concepts/peers.md]: Removed

File: content/concepts/peers.md [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+Conceptually, a Peer Id is a [cryptographic hash][wiki_hash_function] of a peer's
+public key. When peers establish a secure channel, the hash can be used to
+verify that the public key used to secure the channel is the same one used
+to identify the peer.
+
+The [Peer Id spec][spec_peerid] goes into detail about the byte formats used
+for libp2p public keys and how to hash the key to produce a valid Pee Id.
+
+Peer Ids are encoded using the [multihash][definition_multihash] format, which
```
[the remark this answers is not in the document]
Comment (DannyS03) [reply content/concepts/peers.md]: Ah, yes, agreed. Updated

**Review by marten-seemann — APPROVED** [review]
LGTM. We should probably do a `sed s/Peer Id/Peer ID/g` over this document before merging. @DannyS03 feel free to merge the PR once you done that without another round of review.
